### DISCUSSION TITLE
Question on Bundle Rates

### DISCUSSION THREAD

**Discussion by firdaussoberi** [massimo-zaniboni/asterisell-v5#14] [Q&A] [answered]
Hi I got a question on bundle rate as in:
https://www.asterisell.com/manual/bundle-rate-specification-language.html

If we start the bundle without any free minute/second, and do something like this:

> limit-on-first-seconds: 0

does this means the bundle rate is still listed/calculated in the  invoice and the 1st second call is rated accordingly?

And, how would proportional rate works on nested organization hierarchy. Say if we have this:

> limits-are-proportionals-to-activation-date: true

and hierarchy A/B and A/C where A is the billable party (where invoice is generated) and B and C are the two extensions.
B and C follows same bundle rate as A which is say $10, however B was registered for the full month and C was registered on the 15th (50%).

By design, would the total subscription/bundle rate displayed on the invoice to party A be $15 (10+5) then?

Reason is our current setup now prioritizes extensions pbx-style, since a single DID operated (say landline num 123456789) could have multiple extensions based on client criteria. 

Would appreciate your response.
Thanks.

**Comment by massimo-zaniboni** [comment]
> If we start the bundle without any free minute/second, and do something like this:
> limit-on-first-seconds: 0
> does this means the bundle rate is still listed on invoice and the 1st second call is rated accordingly?

The bundle-rate in this case, acts like a service: a CDR with the cost of the bundle-rate is generated (i.e. it is like the cost of a service), but there will be no benefits for the customer, because there will be no calls inside the bundle, i.e. the bundle has a capacity of 0 seconds, and so it starts already full.

**Reply by firdaussoberi** [reply]
Thanks, that clear things up. Just to clarify, on the multiplier based on the hierarchy, if two extensions follow same billable party and bundle rate, the displayed charge should be twice (for full month/proportion) right?

**Reply by massimo-zaniboni** [reply]
Suppose you have ``A``, ``A/B`` and ``A/C``. Every time you assign an explicit price-category associated to a bundle-rate, then the bundle cost will be generated. If you leave the price-category of ``A/B`` null, then it inherits the price-category of ``A``, and no cost will be generated. If you specify a price-category for ``A/C`` then a bundle cost for ``A/C`` will be generated.

If you add a DID to ``A/B``, without specifying a distinct price-category, then no cost will be generated, i.e. the bundle-cost is already the cost of ``A/B``.

In Asterisell ``A/B`` can be an organization with many DID associated, or a specific DID. DID can also be specified using pattern matching rules.

**Reply by firdaussoberi** [reply]
Just to clarify if I understand you correctly. 
When you say this:
>Suppose you have A, A/B and A/C. Every time you assign an explicit price-category associated to a bundle-rate, then the bundle cost will be generated. If you leave the price-category of A/B null, then it inherits the price-category of A, and no cost will be generated. If you specify a price-category for A/C then a bundle cost for A/C will be generated.

1) By no cost, you mean the cost of call outside the coverage of the bundle right? The bundle cost will still be listed right?
Say price-catergory of `A/B` is left null, then `A/B` inherits price-category of `A`. Bundle-cost for `A` is say free 10minutes for $10, and on the 601 second the call cost will be 0 onwards (although a normal rate should be applied)?

2) Or do you mean if `A/B` is left null, although `A/B` inherits price-category of `A`,  bundle-cost will be 0?

**Reply by firdaussoberi** [reply]
I see. I think I finally get what you mean. It is of no cost because it inherits price-category of A under the bundle assuming free call (cost 0), and the fact that now `A` and `B` share the same bundle, which means the bundle charge is once not twice. If price category `A/B` is specified then it uses the limit of `A/B` without using the limit of `A` and two bundle charges are generated 👍

**Reply by firdaussoberi** [reply]
However, I am still as confused. In the features as in
https://www.asterisell.com/pages/features.html#features

The bundle charge shown here looks like it is multiplied by (16) number of lines/extensions:
<img width="418" alt="Screenshot 2022-01-18 at 7 26 11 PM" src="https://user-images.githubusercontent.com/8426299/149929221-72607758-25f4-4fc9-9af7-cfd89cf5fba7.png">

I was looking to replicate this under same billable party and multiple extensions.
If say we need an explicit price category for A/B, A/C, A/D and so on..would say 10 extensions under A would all need 10 different price categories to reflect this multiplier?

Sorry for my ignorance.

**Reply by massimo-zaniboni** [reply]
> Or do you mean if A/B is left null, although A/B inherits price-category of A, bundle-cost will be 0?

yes, this.

If price-category of ``A`` is ``bundle-1000`` and price-category of ``A/B`` is NULL, then the price-category of ``A/B`` is still ``bundle-1000``, but the cost of the bundle will be paid only once from ``A``, and not from ``A/B``. All calls of ``A/B`` will decrease the limits of ``A`` bundle. When there is no more any left limit, then ``A/B`` (and also ``A``) will pay the calls using the logic of normal calls.

**Reply by massimo-zaniboni** [reply]
> I was looking to replicate this under same billable party and multiple extensions. If say we need an explicit price category for A/B, A/C, A/D and so on..would say 10 extensions under A would all need 10 different price categories to reflect this multiplier?

If the bundle is price-category ``bundle-1000``, you need only one price-category.

Then if you want 10 extensions (i.e. DID) with 10 distinct bundles, you create 10 extensions ``A/ext-1``, ``A/ext-2``, ... and specify for each of these extension an explicit price-category ``bundle-1000``. Then you will have 10 bundle costs, and each ``A/ext-1``, ``A/ext-2`` will have its distinct bundle limits.

You can install a DEMO and play with it.

**Comment by massimo-zaniboni** [comment, marked as answer]
```
And, how would proportional rate works on nested organization hierarchy. Say if we have this:

limits-are-proportionals-to-activation-date: true

and hierarchy A/B and A/C where A is the billable party (where invoice is generated) and B and C are the two extensions.
B and C follows same bundle rate as A which is say $10, however B was registered for the full month and C was registered on the 15th (50%).

By design, would the total subscription/bundle rate displayed on the invoice to party A be $15 (10+5) then?
```

You apply a bundle-rate to ``A/B`` only when you assign an explicit price-category to it. The date of assignation of this price-category, is the date of the start of the bundle for ``A/B``. It is not affected by the start of the bundle for ``A``.

If you assign a price-category to ``A``, but not to ``A/B``, then ``A/B`` inherits the price-category (and the bundle) of ``A/B`` and the activation date of ``A/B`` remain the same of ``A``.

```
Reason is our current setup now prioritizes extensions pbx-style, since a single DID operated (say landline num 123456789) could have multiple extensions based on client criteria.
```

If you have a customer ``A`` with many DIDS, maybe you can create different virtual groups for managing different type of bundle activations. All the DIDS assigned directly to ``A`` and without a specific price-category, will use the same bundle of ``A`` (i.e. same activation date, they will decrease the limits of ``A``, and so on). The DIDS assigned directly to ``A`` but with a specific price-category, will generate a bundle-cost and they will have their distinct bundle limits.

Then you can create something like ``A/commercial-office`` if you have a group of DIDS with a distinct bundle.

In any case, when you assign a distinct price-category, it can not anymore use the bundle of the parent organization, but only his own.

**Reply by firdaussoberi** [reply]
I see. So create a bundle hierarchy for DID with distinct price category, and all child inherits bundle A unless distinct price category is specified. 👍

**Reply by firdaussoberi** [reply]
We will try this out first, but I have an instinct that we might need your help later. Will shoot an email if so. Thanks for the great explanation! 😃

**Comment by massimo-zaniboni** [comment]
> We will try this out first, but I have an instinct that we might need your help later. Will shoot an email if so. Thanks for the great explanation! smiley

To be frank, the logic of application of nested bundle-rates and nested-organizations is hard-coded in the rating engine, and I cannot easily change it. It requires the creation of a patched rating-engine, using a different logic.

There is no limit in the way one can figure out a rating-plan, but an application can not be infinitely parametric. At a certain point the parameters became ad-hoc code, or less fantasy on the commercial side.
